pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/funnel/window/FunnelEventsFunctionEvalAggregationFunction.java

```diff
@@ -0,0 +1,502 @@
+    List<int[]> stepBlocks = new ArrayList<>(_numSteps);
+    for (ExpressionContext stepExpression : _stepExpressions) {
+      stepBlocks.add(blockValSetMap.get(stepExpression).getIntValuesSV());
+    }
+    List<Object> extraFieldsBlocks = getExtraFieldsBlocks(blockValSetMap);
+    for (int i = 0; i < length; i++) {
+      int[] groupKeys = groupKeysArray[i];
+      boolean stepFound = false;
+      for (int j = 0; j < _numSteps; j++) {
+        if (stepBlocks.get(j)[i] == 1) {
+          for (int groupKey : groupKeys) {
+            PriorityQueue<FunnelStepEventWithExtraFields> stepEvents =
+                getFunnelStepEvents(groupByResultHolder, groupKey);
+            List<Object> extraFields = extractExtraFields(extraFieldsBlocks, i);
+            stepEvents.add(new FunnelStepEventWithExtraFields(new FunnelStepEvent(timestampBlock[i], j), extraFields));
+          }
+          stepFound = true;
+          break;
+        }
+      }
+      // If the mode is KEEP_ALL and no step is found, add a dummy step event with step -1
+      if (_modes.hasKeepAll() && !stepFound) {
+        for (int groupKey : groupKeys) {
+          PriorityQueue<FunnelStepEventWithExtraFields> stepEvents = getFunnelStepEvents(groupByResultHolder, groupKey);
+          List<Object> extraFields = extractExtraFields(extraFieldsBlocks, i);
+          stepEvents.add(new FunnelStepEventWithExtraFields(new FunnelStepEvent(timestampBlock[i], -1), extraFields));
+        }
+      }
+    }
+  }
+
+  private static PriorityQueue<FunnelStepEventWithExtraFields> getFunnelStepEvents(
+      GroupByResultHolder groupByResultHolder,
+      int groupKey) {
+    PriorityQueue<FunnelStepEventWithExtraFields> stepEvents = groupByResultHolder.getResult(groupKey);
+    if (stepEvents == null) {
+      stepEvents = new PriorityQueue<>();
+      groupByResultHolder.setValueForKey(groupKey, stepEvents);
+    }
+    return stepEvents;
+  }
+
+  @Override
+  public PriorityQueue<FunnelStepEventWithExtraFields> extractAggregationResult(
+      AggregationResultHolder aggregationResultHolder) {
+    return aggregationResultHolder.getResult();
+  }
+
+  @Override
+  public PriorityQueue<FunnelStepEventWithExtraFields> extractGroupByResult(GroupByResultHolder groupByResultHolder,
+      int groupKey) {
+    return groupByResultHolder.getResult(groupKey);
+  }
+
+  @Override
+  public PriorityQueue<FunnelStepEventWithExtraFields> merge(
+      PriorityQueue<FunnelStepEventWithExtraFields> intermediateResult1,
+      PriorityQueue<FunnelStepEventWithExtraFields> intermediateResult2) {
+    if (intermediateResult1 == null) {
+      return intermediateResult2;
+    }
+    if (intermediateResult2 == null) {
+      return intermediateResult1;
+    }
+    intermediateResult1.addAll(intermediateResult2);
+    return intermediateResult1;
+  }
+
+  @Override
+  public DataSchema.ColumnDataType getIntermediateResultColumnType() {
+    return DataSchema.ColumnDataType.STRING_ARRAY;
```

Comment:
This should be `OBJECT`, and you may override the ser/de method. You shouldn't need to modify other classes for ser/de purpose

Reply:
Done